packages/server/src/utilities/rowProcessor/bbReferenceProcessor.ts

```diff
@@ -1,93 +1,190 @@
+          if (notFoundIds?.length) {
+            throw new InvalidBBRefError(notFoundIds[0], FieldSubtype.USER)
+          }
+
+          if (!referenceIds?.length) {
+            return null
+          }
+
+          if (subtype === FieldSubtype.USERS) {
+            return referenceIds
+          }
+
+          return referenceIds.join(",")
+        }
+        default:
+          throw utils.unreachable(subtype)
       }
+    }
+    case FieldType.BB_REFERENCE_SINGLE: {
+      if (value && Array.isArray(value)) {
+        throw "BB_REFERENCE_SINGLE cannot be an array"
+      }
+
+      const id = typeof value === "string" ? value : value._id
+
+      const user = await cache.user.getUser(id)
 
-      if (subtype === FieldSubtype.USERS) {
-        return referenceIds
+      if (!user) {
+        throw new InvalidBBRefError(id, FieldSubtype.USER)
       }
 
-      return referenceIds.join(",") || null
+      return user._id!
     }
+
     default:
-      throw utils.unreachable(subtype)
+      throw utils.unreachable(type)
   }
 }
 
+interface UserReferenceInfo {
+  _id: string
+  primaryDisplay: string
+  email: string
+  firstName: string
+  lastName: string
+}
+
+export function processOutputBBReferences(
+  value: string,
+  type: FieldType.BB_REFERENCE_SINGLE
+): Promise<UserReferenceInfo>
+export function processOutputBBReferences(
+  value: string,
+  type: FieldType.BB_REFERENCE,
+  subtype: FieldSubtype.USER | FieldSubtype.USERS
+): Promise<UserReferenceInfo[]>
+
 export async function processOutputBBReferences(
   value: string | string[],
-  subtype: FieldSubtype.USER | FieldSubtype.USERS
+  type: FieldType.BB_REFERENCE | FieldType.BB_REFERENCE_SINGLE,
+  subtype?: FieldSubtype
 ) {
   if (value === null || value === undefined) {
     // Already processed or nothing to process
     return value || undefined
   }
 
-  const ids =
-    typeof value === "string" ? value.split(",").filter(id => !!id) : value
+  switch (type) {
+    case FieldType.BB_REFERENCE: {
+      const ids =
+        typeof value === "string" ? value.split(",").filter(id => !!id) : value
 
-  switch (subtype) {
-    case FieldSubtype.USER:
-    case FieldSubtype.USERS: {
-      const { users } = await cache.user.getUsers(ids)
-      if (!users.length) {
+      switch (subtype) {
+        case undefined:
+          throw "Subtype must be defined"
+        case FieldSubtype.USER:
+        case FieldSubtype.USERS: {
+          const { users } = await cache.user.getUsers(ids)
+          if (!users.length) {
+            return undefined
+          }
+
+          return users.map(u => ({
+            _id: u._id,
+            primaryDisplay: u.email,
+            email: u.email,
+            firstName: u.firstName,
+            lastName: u.lastName,
+          }))
+        }
+        default:
+          throw utils.unreachable(subtype)
+      }
+    }
+    case FieldType.BB_REFERENCE_SINGLE: {
+      if (!value) {
```

Comment:
Should this uncertainty be reflected in the function signature? Right now there's no indication that `value` could be undefined, or the function can return undefined.